### DISCUSSION TITLE
Speed up commands to sub devices

### DISCUSSION THREAD

**Discussion by nospam2k** [jasonacox/tinytuya#533] [Q&A]
I have 40 disk lights connecting evenly among 3 bt gateways. Using the app, I can send a group command that turns off all lights simultaneously. I am trying to code the same thing but the best I've been able to accomplish is the following which still takes about 3 seconds to run. Is there a group command or a way to send to every device on a gateway, etc.:

```python
import socket
import json
import tinytuya
import time
import threading

tinytuya.set_debug(False)

with open('endpoints.json', 'r') as endpoints_file:
  endpoints = json.load(endpoints_file)

gws = {}

for i in range(40):

  dev_name = list(endpoints.keys())[i];
  gw_id = list(endpoints.values())[i].get('tuya_gwid')
  gw_ip = list(endpoints.values())[i].get('tuya_ip')
  gw_key = list(endpoints.values())[i].get('tuya_key')
  dev_id = list(endpoints.values())[i].get('tuya_id')
  dev_cid = list(endpoints.values())[i].get('tuya_cid')
  dev_ver = list(endpoints.values())[i].get('tuya_version')

  if gw_ip not in gws:
    gws[gw_ip] = {"gw": tinytuya.Device(gw_id, gw_ip, gw_key, version=3.3), "dev": []}

  dev_ids = [device["dev_id"] for device in gws[gw_ip]["dev"]]
  if dev_id not in dev_ids:
      gws[gw_ip]["dev"].append({"dev_id": dev_id, "dev_cid": dev_cid})

def send_commands(gw_ip, gw_data):
  gw = gw_data["gw"]
  gw.set_socketTimeout(10)

  try:        
    for device in gw_data["dev"]:
      dev_id = device["dev_id"]
      dev_cid = device["dev_cid"]

      print(f"Sending command to device {dev_id} at gateway {gw_ip}")
      d = tinytuya.Device(dev_id, gw_ip, version=3.3, cid=dev_cid, parent=gw)
      d.set_multiple_values({"1": False}, nowait=True)
      #time.sleep(.1)

  except Exception as e:
    print(f"Failed to send command to device {dev_id} at gateway {gw_ip}: {e}")

threads = []

for gw_ip, gw_data in gws.items():
  thread = threading.Thread(target=send_commands, args=(gw_ip, gw_data))
  threads.append(thread)
  thread.start()

for thread in threads:
  thread.join()

print("All commands sent successfully.")
```

**Comment by jasonacox** [comment]
I see you are using `nowait=True` which is good, but would also try set_socketPersistent(True) so that you avoid the TCP setup/teardown delay. However, that would only be helpful if you plan on having your script send multiple commands to the device (eg. a service that stays running and keeps persistent connections vs. run once script).

**Comment by uzlonewolf** [comment]
I don't know of any group commands, but if you can find one I would be interested.

> ```python
> gws[gw_ip] = {"gw": tinytuya.Device(gw_id, gw_ip, gw_key, version=3.3), "dev": []}
> ```

The issue with the above is tinytuya makes a connection to the gateway, sends the command to 1 light, and then disconnects.  It then repeats this for all 40 lights.  Switching to persistent mode by either adding `persist=True` to the above line or calling `.set_socketPersistent(True)` should reduce the total time to switch all 40 lights greatly.
